File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/catalogue/Product.kt
package nl.zappymart.domain.catalogue

import nl.zappymart.domain.shared.Money

data class Product(
    val id: String,
    val name: String,
    val slug: String,
    val description: String,
    val price: Money,
    val category: Category,
    val stock: Int,
    val imageUrl: String?,
) {

    fun hasStockFor(quantity: Int): Boolean = stock >= quantity

    fun withStockReducedBy(quantity: Int): Product {
        require(hasStockFor(quantity)) { "Product $id has $stock in stock and cannot give up $quantity" }
        return copy(stock = stock - quantity)
    }
}

File: backends/node/subgraphs/catalogue/src/domain/product.ts
import type { Money } from "@zappy/shared";

export type Product = {
  readonly id: string;
  readonly name: string;
  readonly slug: string;
  readonly description: string;
  readonly price: Money;
  readonly categorySlug: string;
  readonly stock: number;
  readonly imageUrl: string | null;
};

export type Category = {
  readonly id: string;
  readonly name: string;
  readonly slug: string;
};

export function hasStockFor(product: Product, quantity: number): boolean {
  return quantity >= 1 && product.stock >= quantity;
}

export function withStockReducedBy(product: Product, quantity: number): Product {
  if (!hasStockFor(product, quantity)) {
    throw new Error(`Product ${product.id} has ${product.stock} in stock and cannot give ${quantity}`);
  }
  return { ...product, stock: product.stock - quantity };
}

export function withStockRaisedBy(product: Product, quantity: number): Product {
  return { ...product, stock: product.stock + quantity };
}
